astropy/io/misc/parquet.py

```diff
@@ -507,3 +507,182 @@ def get_pyarrow():
     from pyarrow import parquet
 
     return pa, parquet
+
+
+def write_parquet_votable(
+    table, output, *, metadata=None, overwrite=False, overwrite_metadata=False
+):
+    """
+    Writes a Parquet file with a VOT (XML) metadata table included.
+
+    Parameters
+    ----------
+    table : `~astropy.table.Table`
+        Data table that is to be written to output.
+    output : str or path-like
+        The filename to write the table to.
+    metadata : dict
+        Nested dictionary (keys = column names; sub-keys = meta keys) for each
+        of the columns containing a dictionary with metadata. Existing metadata
+        takes precedent, use ``overwrite_metadata`` to ensure this dictionary is
+        being used in all cases.
+    overwrite : bool, optional
+        If `True`, overwrite the output file if it exists. Raises an
+        ``OSError`` if ``False`` and the output file exists. Default is `False`.
+    overwrite_metadata : bool, optional
+        If `True`, overwrite existing column metadata. Default is `False`.
+    """
+    # TODO cases to handle:
+    # - overwriting metadata, metadata could be partially missing, the provided
+    #   one then could overwrite the existing one in the table
+    # - make overwrite actually overwrite rather than delete the file upfront
+    # - warn for non VO standard units at write, not just at read time
+    # - deal better with non-VO units
+
+    import io
+    import xml.etree.ElementTree
+
+    import pyarrow.parquet
+
+    from astropy.io.votable.tree import VOTableFile
+
+    if not isinstance(output, (str, os.PathLike)):
+        raise TypeError(f"`output` should be a string or path-like, not {output}")
+    output = Path(output)
+
+    if Path.exists(output):
+        if overwrite:
+            # We must remove the file prior to writing below.
+            Path.unlink(output)
+        else:
+            raise OSError(NOT_OVERWRITING_MSG.format(output))
+
+    # Prepare the VOTable (XML)
+    # We only use the first row of the astropy table to get the general
+    # information such as arraysize, ID, or datatype.
+
+    # TODO this step looses the metadata that the astropy Table input might have had,
+    # e.g. column units.
+    votablefile = VOTableFile()
+    votable_write = votablefile.from_table(table[0:1])
+
+    # TODO: API placeholder for inheriting metadata from existing table, thus
+    # no API change is needed for making this technically optional
+    if metadata is None:
+        raise NotImplementedError("metadata has to be always specified")
+
+    # Then add the other metadata keys to the FIELDS parameters of the VOTable
+    metadatakeys = list(metadata[next(iter(metadata.keys()))].keys())
```

Comment:
I may be misreading this, but it looks like `metadatakeys` will just be the attributes supplied for the first column name returned by `keys()`.  In the loop below that list of attributes is applied to all fields, so this seems to assume that every column in `metadata` would have the same attributes.  

So for example if the first column didn't have a `description`, then none of the other column's descriptions would get added.  If the first column had `description` but another column didn't, then there would be a key errors assigning `field.description = metadata[field.name]["description"]` for that field.

Reply:
If the above is a problem, I'd be tempted instead to loop over the `metadata` keys, find the relevant field with `get_field_by_id_or_name()`, then apply the attributes.

This would only break in the edge cases I mentioned, so I think it's fine to address this along with the other metadata TODOs mentioned in the code.